TransformerEngine Tests failure: Assertion failed: A_scale_inverse.is_cuda() && B_scale_inverse.is_cuda(). Scaling factors must be on device.

Post #1615, TE tests are failing

Example Failure
```python
____________________________________________________________________________________________________________________________ test_te_linear_forward_backward_multiple_iteration[fp8_recipe1] _____________________________________________________________________________________________________________________________

fp8_recipe = margin=0, format=HYBRID, amax_history_len=1024, wgrad_override=False, fp8_dpa=False, fp8_mha=False

    @requiresCUDA
    @pytest.mark.parametrize("fp8_recipe", recipes)
    def test_te_linear_forward_backward_multiple_iteration(fp8_recipe):
        # Test Description:
        # In this test, we verify whether a model using TransformerEngine Linear
        # and transformer_engine executor converge to same state.
        # Since, the FP8 operations are stateful, we want to verify that
        # our output matches over multiple iterations (where state handling comes into picture)
        dtype = torch.bfloat16
        device = "cuda"
        iterations = 5
    
        # TE inputs
        input_shape = (768, 4096)
        te_linear1 = te.Linear(4096, 4096, params_dtype=dtype)
        te_linear2 = te.Linear(4096, 2048, params_dtype=dtype)
    
        def clone_params(*params):
            return tuple(param.detach().clone() for param in params)
    
        # Parameters for thunder to optimize
        w1, w2, b1, b2 = clone_params(te_linear1.weight, te_linear2.weight, te_linear1.bias, te_linear2.bias)
    
        target_value = torch.tensor(42, dtype=dtype, device=device)
    
        def train_model(model, optimizer):
            # Run for `iterations`.
            for iter_n in range(iterations):
                x = torch.ones(*input_shape, device=device, dtype=dtype) + iter_n
                result = model(x)
                loss = torch.nn.functional.mse_loss(result.sum(), target_value)
                loss.backward()
                optimizer.step()
                optimizer.zero_grad()
    
        def te_model(x):
            # Enable autocasting for the forward pass
            with te.fp8_autocast(fp8_recipe=fp8_recipe):
                return te_linear2(te_linear1(x))
    
        te_sgd_optimizer = torch.optim.SGD(list(te_linear1.parameters()) + list(te_linear2.parameters()))
    
        train_model(te_model, te_sgd_optimizer)
    
        def fn(x, w1, w2, b1, b2):
            o = torch.nn.functional.linear(x, w1, b1)
            return torch.nn.functional.linear(o, w2, b2)
    
        cfn = thunder.jit(fn, executors=[transformer_engine_ex])
    
        # Enable grad on thunder params.
        list(map(lambda t: t.requires_grad_(True), (w1, w2, b1, b2)))
        thunder_sgd_optimizer = torch.optim.SGD([w1, w2, b1, b2])
    
        def thunder_model(x):
            return cfn(x, w1, w2, b1, b2)
    
>       train_model(thunder_model, thunder_sgd_optimizer)

thunder/tests/test_transformer_engine_executor.py:138: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
thunder/tests/test_transformer_engine_executor.py:112: in train_model
    loss.backward()
/usr/local/lib/python3.12/dist-packages/torch/_tensor.py:626: in backward
    torch.autograd.backward(
/usr/local/lib/python3.12/dist-packages/torch/autograd/__init__.py:347: in backward
    _engine_run_backward(
/usr/local/lib/python3.12/dist-packages/torch/autograd/graph.py:823: in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/usr/local/lib/python3.12/dist-packages/torch/autograd/function.py:307: in apply
    return user_fn(self, *args)
/usr/local/lib/python3.12/dist-packages/torch/autograd/function.py:600: in wrapper
    outputs = fn(ctx, *args)
thunder/executors/torch_autograd.py:144: in backward
    grads = ctx.compiled_backward([saved_tensors_list, ctx.saved_other], args)
/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py:116: in decorate_context
    return func(*args, **kwargs)
thunder/executors/torchex.py:178: in no_autocast_fn
    return fn(*args, **kwargs)
thunder.backward_fn_14:27: in backward_fn
    (t24, t25, t26) = te_functional_linear_backward((768, 4096), (2048, 4096), (2048,), ctx_te_14, (t14, t15, t16, t17, t18, None), t20)
thunder/executors/transformer_engineex.py:384: in _te_functional_linear_backward_impl
    grads = _Linear.backward(ctx, g)
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/module/linear.py:533: in backward
    wgrad, _ = fp8_gemm(
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/cpp_extensions/gemm.py:212: in fp8_gemm
    _ = fn(*args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <OpOverloadPacket(op='tex_ts.te_gemm_ts')>, args = (tensor([[ 20,  24,  20,  ...,  24,  20,  24],
        [192,  62, 192,  ...,  62, 192,  62],
        [141, 195, 141,  ...,  63,  63,  63]], device='cuda:0',
       dtype=torch.uint8), None, 0, <DType.kFloat8E4M3: 6>, [-1, -1, 1], True, ...), kwargs = {}

    def __call__(self, /, *args, **kwargs):
        # overloading __call__ to ensure torch.ops.foo.bar()
        # is still callable from JIT
        # We save the function ptr as the `op` attribute on
        # OpOverloadPacket to access it here.
    
        # Directly calling OverloadPacket goes into C++, which will check
        # the schema and cause an error for torchbind op when inputs consist of FakeScriptObject so we
        # intercept it here and call TorchBindOpverload instead.
        if self._has_torchbind_op_overload and _must_dispatch_in_python(args, kwargs):
            return _call_overload_packet_from_python(self, args, kwargs)
>       return self._op(*args, **(kwargs or {}))
E       RuntimeError: /workspace/transformerengine/transformer_engine/pytorch/csrc/ts_fp8_op.cpp:260 in function te_gemm_ts: Assertion failed: A_scale_inverse.is_cuda() && B_scale_inverse.is_cuda(). Scaling factors must be on device.

/usr/local/lib/python3.12/dist-packages/torch/_ops.py:1123: RuntimeError
```

All Failures
```
================================================================================================================================================ short test summary info =================================================================================================================================================
FAILED thunder/tests/test_transformer_engine_executor.py::test_te_linear_forward_backward[None] - RuntimeError: /workspace/transformerengine/transformer_engine/pytorch/csrc/ts_fp8_op.cpp:260 in function te_gemm_ts: Assertion failed: A_scale_inverse.is_cuda() && B_scale_inverse.is_cuda(). Scaling factors must be on device.
FAILED thunder/tests/test_transformer_engine_executor.py::test_te_linear_forward_backward[fp8_recipe1] - RuntimeError: /workspace/transformerengine/transformer_engine/pytorch/csrc/ts_fp8_op.cpp:260 in function te_gemm_ts: Assertion failed: A_scale_inverse.is_cuda() && B_scale_inverse.is_cuda(). Scaling factors must be on device.
FAILED thunder/tests/test_transformer_engine_executor.py::test_te_linear_forward_backward_multiple_iteration[None] - RuntimeError: /workspace/transformerengine/transformer_engine/pytorch/csrc/ts_fp8_op.cpp:260 in function te_gemm_ts: Assertion failed: A_scale_inverse.is_cuda() && B_scale_inverse.is_cuda(). Scaling factors must be on device.
FAILED thunder/tests/test_transformer_engine_executor.py::test_te_linear_forward_backward_multiple_iteration[fp8_recipe1] - RuntimeError: /workspace/transformerengine/transformer_engine/pytorch/csrc/ts_fp8_op.cpp:260 in function te_gemm_ts: Assertion failed: A_scale_inverse.is_cuda() && B_scale_inverse.is_cuda(). Scaling factors must be on device.
FAILED thunder/tests/test_transformer_engine_executor.py::test_te_with_retain_graph - RuntimeError: /workspace/transformerengine/transformer_engine/pytorch/csrc/ts_fp8_op.cpp:260 in function te_gemm_ts: Assertion failed: A_scale_inverse.is_cuda() && B_scale_inverse.is_cuda(). Scaling factors must be on device.
======================================================================================================================================== 5 failed, 3 passed, 34 warnings in 3.92s ========================================================================================================================================
```

--- a/thunder/core/prims.py
+++ b/thunder/core/prims.py
@@ -291,6 +291,8 @@ class OpTags(Enum):
     AUTO_REGISTERED = auto()
     # Label for operations representing enter/exit of context managers.
     CTX_MANAGER_ENTER_EXIT_OP = auto()
+    # Label to explicitly disable an operation from recomputing in backward - see function `recompute_saved_for_backward`.
+    DONT_RECOMPUTE_IN_BACKWARD = auto()
 
 
 # TODO RC1 Document this function and describe the parts of a primitive

--- a/thunder/core/transforms.py
+++ b/thunder/core/transforms.py
@@ -3220,7 +3220,9 @@ def recompute_saved_for_backward(fwd_trace: Trace, bwd_trace: Trace) -> tuple[Tr
 
             proxy_names_to_producers[o.name] = bsym
             all_proxies.add(vo)
-            if ProxyTag.RECOMPUTE_IN_BACKWARD in o.tags and not has_tags(bsym, {prims.OpTags.RANDOM_OP}):
+            if ProxyTag.RECOMPUTE_IN_BACKWARD in o.tags and not has_tags(
+                bsym, {prims.OpTags.RANDOM_OP, prims.OpTags.DONT_RECOMPUTE_IN_BACKWARD}
+            ):
                 all_recomputable_proxies.add(vo)
 
     rematerializable = old_saved_for_bwd & all_recomputable_proxies

--- a/thunder/executors/transformer_engineex.py
+++ b/thunder/executors/transformer_engineex.py
@@ -438,7 +438,12 @@ def _create_fp8_linear_bound_symbol(
 
     meta_fn = make_te_linear_meta(is_grad_enabled=is_grad_enabled)
     sym = Symbol(
-        name=name, meta=meta_fn, is_prim=True, executor=transformer_engine_ex, _bind_postprocess=bind_postprocess
+        name=name,
+        meta=meta_fn,
+        is_prim=True,
+        executor=transformer_engine_ex,
+        _bind_postprocess=bind_postprocess,
+        tags=(prims.OpTags.DONT_RECOMPUTE_IN_BACKWARD,),
     )
     bsym = sym.bind(a, w, b, output=meta_fn(a, w, b))
 

